#include "managerreservewidget.h"
#include "ui_managerreservewidget.h"

ManagerReserveWidget::ManagerReserveWidget(ReserveStruct rstruct,
                                           QWidget *parent) :
    QWidget(parent),
    ui(new Ui::ManagerReserveWidget)
{
    ui->setupUi(this);
    ui->table_widget->setSelectionMode(QTableView::ExtendedSelection);
    QModelIndex cur = rstruct.selection_model->currentIndex();
    max_docs = 10;
    oldmodel = 0;
    grey_table = 0;
    this->source = rstruct.source_model;
    this->list_model = rstruct.list_model;
    this->selection_model = rstruct.selection_model;
    this->table_header = rstruct.table_header;
    //this->table_header->setParent(ui->table_widget);
    //QHeaderView *table_headerd = new QHeaderView(Qt::Horizontal);
    //ui->table_widget->setHorizontalHeader(table_header);
    //ui->table_widget->horizontalHeader()->setVisible(true);
    ui->table_widget->horizontalHeader()->setHighlightSections(false);

    set_columns_names();
    set_layout();

    fill_list();
    if(cur.isValid())
        this->ui->docs_list->setCurrentIndex(cur);

    connects();
}

ManagerReserveWidget::~ManagerReserveWidget()
{
    delete ui;
}

void ManagerReserveWidget::connects(){
    connect(ui->docs_list, SIGNAL(clicked(QModelIndex)), SLOT(change_doc(QModelIndex)));
    connect(ui->docs_list->selectionModel(), SIGNAL(currentChanged(QModelIndex, QModelIndex)), SLOT(change_doc(QModelIndex, QModelIndex)));
    connect(ui->plus_button, SIGNAL(clicked()), SLOT(add_position()));
    connect(ui->minus_button, SIGNAL(clicked()), SLOT(subtr_quant()));
    connect(ui->delete_button, SIGNAL(clicked()), SLOT(delete_position()));
    //connect(ui->show_button, SIGNAL(clicked()), SLOT(view_doc()));
    connect(ui->show_button, SIGNAL(pressed()), SLOT(view_doc()));
    connect(ui->show_button, SIGNAL(released()), SLOT(restore_doc()));
    connect(ui->clear_button, SIGNAL(clicked()), SLOT(clear_doc()));
    connect(ui->close_button, SIGNAL(clicked()), SIGNAL(switch_reserves()));
    connect(ui->move_button, SIGNAL(clicked()), SLOT(move_doc()));
}

void ManagerReserveWidget::set_layout(){
    // —Å–æ—Ö—Ä–∞–Ω—è–µ–º —Ä–∞–∑–º–µ—Ä —à—Ä–∏—Ñ—Ç–∞, —á—Ç–æ–±—ã –Ω–∞ –µ–≥–æ –æ—Å–Ω–æ–≤–µ —Ä–µ—Å–∞–π–∑–∏—Ç—å –≤–∏–¥–∂–µ—Ç—ã
    int fsize = qApp->font().pointSize();

    ui->plus_button->setFixedSize(fsize*2.5, fsize*5.5);
    ui->minus_button->setFixedSize(fsize*2.5, fsize*2.5);
    ui->delete_button->setFixedSize(ui->minus_button->size());

    ui->show_button->setFixedSize(fsize*10, fsize*2.5);
    ui->clear_button->setFixedSize(ui->show_button->size());
    ui->move_button->setFixedSize(ui->show_button->size());
    ui->close_button->setFixedSize(ui->show_button->size());
    ui->tobase_button->setFixedSize(ui->plus_button->height(), ui->plus_button->height());

    ui->note_button->setFixedSize(fsize*2.5, fsize*2.5);

    ui->docs_list->setFixedHeight(fsize*16);
    ui->docs_list->setFixedWidth(fsize*10);
    ui->note_edit->setFixedHeight(fsize*2.5);
    //ui->table_widget->setFixedHeight(ui->docs_list->height() - ui->note_edit->height());
    //ui->right_frame->setFixedHeight(ui->docs_list->height());
    ui->main_frame->setFixedHeight(ui->docs_list->height());

    ui->table_widget->verticalHeader()->setDefaultSectionSize(qApp->font().pointSize()*2);

    this->setFixedSize(qApp->desktop()->maximumWidth(), ui->main_frame->frameSize().height());
}

void ManagerReserveWidget::set_grey_table(GreyTable *g){
    this->grey_table = g;
}

int ManagerReserveWidget::resize_section(int num, int w){
    int old = ui->table_widget->horizontalHeader()->sectionSize(num);
    ui->table_widget->horizontalHeader()->resizeSection(num, w);
    return old;
}

void ManagerReserveWidget::set_reserve_header(){
    this->ui->table_widget->setHorizontalHeader(table_header);
    //this->table_header->setParent(ui->table_widget);
    this->table_header->show();
}

void ManagerReserveWidget::fill_list(){
    /*for(int i = 1; i <= max_docs; i++){
        QListWidgetItem *ql = new QListWidgetItem(QString("–ó–∞—è–≤–∫–∞ %1").arg(i), ui->docs_list);
        ql->setData(Qt::UserRole, i);
    }
*/
    ui->docs_list->setModel(list_model);
    ui->docs_list->setSelectionModel(selection_model);
    ui->docs_list->setCurrentIndex(list_model->index(0, 0));
    change_doc(list_model->index(0, 0));
}

void ManagerReserveWidget::set_columns_names(){
    columns_names["id"] = "";
    columns_names["grey_id"] = "id";
    columns_names["name"] = "–ù–∞–∑–≤–∞–Ω–∏–µ";
    columns_names["quantity"] = "–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ";
    columns_names["price"] = "–¶–µ–Ω–∞";
    columns_names["sumprice"] = "–°—É–º–º–∞";
    columns_names["year"] = "–ì–æ–¥";
    columns_names["storage"] = "–°–∫–ª–∞–¥";
    columns_names["rack"] = "–°—Ç–µ–ª–ª–∞–∂";
    columns_names["board"] = "–ü–æ–ª–∫–∞";
    columns_names["box"] = "–Ø—â–∏–∫";
    columns_names["note"] = "–¢—Ä–µ–±–æ–≤–∞–Ω–∏–µ";
}

void ManagerReserveWidget::set_headers(){
    QString hdr;
    for(int i = 0; i < model->columnCount(); i++){
        //qDebug() << model->headerData(i, Qt::Horizontal).toString();
        hdr = columns_names[model->headerData(i, Qt::Horizontal).toString()];
        if(hdr != QString())
            model->setHeaderData(i, Qt::Horizontal, hdr);
    }
}

void ManagerReserveWidget::change_doc(QModelIndex cur, QModelIndex prev){
    //set_reserve_header();
    ui->num_label->setText(QString("–ó–∞—è–≤–∫–∞ %1").arg(list_model->data(cur, Qt::UserRole).toString()));
    // –µ—Å–ª–∏ –≤ –±–∞–∑–µ –Ω–µ—Ç —Ç–∞–∫–æ–≥–æ –¥–æ–∫—É–º–µ–Ω—Ç–∞ (–Ω—É –ø—Ä–æ–≥—Ä–∞–º–º–∞ –ø–µ—Ä–≤—ã–π —Ä–∞–∑ –∑–∞–ø—É—Å–∫–∞–µ—Ç—Å—è) - —Å–æ–∑–¥–∞—ë–º –µ–≥–æ –≤ reserve_docs
    // –∞ –≤–æ–æ–±—â–µ, —ç—Ç–æ —Å—Ç—ã–¥ –∏ —Å—Ä–∞–º, —É–±—Ä–∞—Ç—å –≤–ø–æ—Å–ª–µ–¥—Å—Ç–≤–∏–∏
    QSqlQuery query(base);
    QString query_str = QString("SELECT count(*) FROM reserve_docs WHERE doc_num = %1 AND user_id = %2").arg(cur.data(Qt::UserRole).toInt()).arg(USER_ID);
    if(!query.exec(query_str)){
        QMessageBox::information(this, "–í—Å—ë –ø–ª–æ—Ö–æ", query.lastError().text(), QMessageBox::Ok);
        return;
    }
    query.next();
    if(query.value(0).toInt() == 0){
        // —Ç—É—Ç-—Ç–æ –∏ —Å–æ–∑–¥–∞—ë–º –µ–≥–æ
        query_str = QString("INSERT INTO reserve_docs (user_id, doc_num) VALUES (%1, %2)").arg(USER_ID).arg(cur.data(Qt::UserRole).toInt());
        if(!query.exec(query_str)){
            QMessageBox::information(this, "–í—Å—ë —Å–Ω–æ–≤–∞ –ø–ª–æ—Ö–æ", query.lastError().text(), QMessageBox::Ok);
            return;
        }
    }
    // –ø–æ—Å–ª–µ - –ø–æ–∫–∞–∑—ã–≤–∞–µ–º —Å–æ–∂–µ—Ä–∂–∏–º–æ–µ
    delete ui->table_widget->model();
    query_str = QString("SELECT r.id as id, "
                        "r.grey_id as grey_id,  "
                        "t.name as name, "
                        "r.quantity as quantity, "
                        "r.price as price, "
                        "(r.price*r.quantity) as sumprice, "
                        "g.year as year, "
                        "p.storage as storage, "
                        "p.rack as rack, "
                        "p.board as board, "
                        "p.box as box, "
                        "r.note as note "
                        "FROM reserve r JOIN greytable g ON r.grey_id = g.id JOIN trademarks t ON g.trademark_id = t.id JOIN places p ON g.place_id = p.id "
                        "WHERE r.doc_id = (SELECT id FROM reserve_docs WHERE user_id = %1 AND doc_num = %2)")
            .arg(USER_ID)
            .arg(cur.data(Qt::UserRole).toInt());
    //qDebug() << query_str;
    model = new QSortFilterProxyModel(this);
    source->setQuery(query_str, base);
    if(source->lastError().isValid()){
        QMessageBox::information(this, "–ù–µ —É–¥–∞–ª–æ—Å—å —Å—Ñ–æ—Ä–º–∏—Ä–æ–≤–∞—Ç—å —Ç–∞–±–ª–∏—Ü—É", source->lastError().text(), QMessageBox::Ok);
        return;
    }
    model->setSourceModel(source);
    set_headers();
    ui->table_widget->setModel(model);
    ui->table_widget->hideColumn(0);
    ui->table_widget->resizeColumnsToContents();
    //ui->table_widget->resizeRowsToContents();

    double sum = count_sum();
    list_model->setData(ui->docs_list->currentIndex(), (sum > 0 ? Qt::red : Qt::black), Qt::TextColorRole);
}

void ManagerReserveWidget::add_position(){
    if(GLOBAL_MODE != SEARCHER_GREY_MODE)
        return;

    int grey_id = grey_table->table_data(columns_grey_ids["id"]).toInt();
    int doc_num = ui->docs_list->currentIndex().data(Qt::UserRole).toInt();
    int quantity = grey_table->table_data(columns_grey_ids["quantity"]).toInt();
    int retail_price = grey_table->table_data(columns_grey_ids["price_ret"]).toDouble();
    int whole_price = grey_table->table_data(columns_grey_ids["price_whole"]).toDouble();
    int whole_begin = grey_table->table_data(columns_grey_ids["whole_begin"]).toInt();

    ReserveAddPosition *add = new ReserveAddPosition(grey_id, doc_num, quantity, retail_price, whole_price, whole_begin, this);
    if(add->exec() == QDialog::Accepted){
        //refresh_table();
        change_doc(ui->docs_list->currentIndex());
    }

    emit need_refresh();
}

void ManagerReserveWidget::refresh_table(){
    if(GLOBAL_MODE != SEARCHER_GREY_MODE)
        return;
    grey_table->refresh_table();
}

double ManagerReserveWidget::count_sum(){
    double sum = 0;
    for(int i = 0; i < ui->table_widget->model()->rowCount(); i++)
        sum += ui->table_widget->model()->data(ui->table_widget->model()->index(i, 5)).toDouble();
    ui->summ_label->setText(QString::number(sum, 'f', 2));
    return sum;
}

void ManagerReserveWidget::subtr_quant(){
    if(ui->table_widget->selectionModel()->selectedRows().count() > 1)
        return;
    int id = ui->table_widget->model()->data(ui->table_widget->model()->index(ui->table_widget->currentIndex().row(), 0)).toInt();
    int quantity = ui->table_widget->model()->data(ui->table_widget->model()->index(ui->table_widget->currentIndex().row(), 3)).toInt();
    ReserveSubtract *rs = new ReserveSubtract(id, quantity, this);
    if(rs->exec() == QDialog::Accepted){
        //refresh_table();
        change_doc(ui->docs_list->currentIndex());
    }
    emit need_refresh();
}

void ManagerReserveWidget::delete_position(){
    // —É–¥–∞–ª–∏—Ç—å –ø–æ–∑–∏—Ü–∏—é –∏–∑ —Ç–∞–±–ª–∏—Ü—ã –∏ –≤–µ—Ä–Ω—É—Ç—å –≤ –±–∞–∑—É
    if(QMessageBox::question(this, "–£–¥–∞–ª–∏—Ç—å –ø–æ–∑–∏—Ü–∏–∏?", "–£–¥–∞–ª–∏—Ç—å –≤—ã–±—Ä–∞–Ω–Ω—ã–µ –ø–æ–∑–∏—Ü–∏–∏?", QMessageBox::Yes, QMessageBox::No) == QMessageBox::Yes){
        foreach(QModelIndex i, ui->table_widget->selectionModel()->selectedRows()){
            int id = ui->table_widget->model()->data(model->index(i.row(), 0)).toInt();
            delpos(id);
        }
        //refresh_table();
        change_doc(ui->docs_list->currentIndex());
        emit need_refresh();
    }
}

void ManagerReserveWidget::delpos(int id){
    QSqlQuery query(base);
    QString query_str = QString("DELETE FROM reserve WHERE id = %1").arg(id);
    if(!query.exec(query_str)){
        QMessageBox::warning(this, "–ù–µ —É–¥–∞–ª–æ—Å—å —É–¥–∞–ª–∏—Ç—å", query.lastError().text(), QMessageBox::Ok);
        return;
    }
    //refresh_table();
}

void ManagerReserveWidget::clear_doc(){
    if(QMessageBox::question(this, "–û—á–∏—Å—Ç–∏—Ç—å –∑–∞—è–≤–∫—É?", "–û—á–∏—Å—Ç–∏—Ç—å –∑–∞—è–≤–∫—É –∏ –≤–µ—Ä–Ω—É—Ç—å –≤—Å—ë –≤ –±–∞–∑—É?", QMessageBox::Yes, QMessageBox::No) == QMessageBox::Yes){
        int c = 0;
        while(c < ui->table_widget->model()->rowCount()){
            int id = ui->table_widget->model()->data(ui->table_widget->model()->index(c, 0)).toInt();
            delpos(id);
            c++;
        }
    }
    change_doc(ui->docs_list->currentIndex());
    emit need_refresh();
}

void ManagerReserveWidget::move_doc(){


}

void ManagerReserveWidget::view_doc(){
    oldmodel = static_cast<QSortFilterProxyModel*>(ui->table_widget->model());

    QString query_str = QString("SELECT t.name, sum(r.quantity) "
                        "FROM reserve r JOIN greytable g ON r.grey_id = g.id JOIN trademarks t ON g.trademark_id = t.id "
                        "WHERE r.doc_id = (SELECT id FROM reserve_docs WHERE user_id = %1 AND doc_num = %2) GROUP BY t.name")
            .arg(USER_ID)
            .arg(ui->docs_list->currentIndex().data(Qt::UserRole).toInt());
    //qDebug() << query_str;
    QSqlQueryModel *model = new QSqlQueryModel(this);
    model->setQuery(query_str, base);
    if(model->lastError().isValid()){
        QMessageBox::information(this, "–ù–µ —É–¥–∞–ª–æ—Å—å —Å—Ñ–æ—Ä–º–∏—Ä–æ–≤–∞—Ç—å —Ç–∞–±–ª–∏—Ü—É", model->lastError().text(), QMessageBox::Ok);
        return;
    }
    ui->table_widget->setModel(model);
    ui->table_widget->resizeColumnsToContents();

}

void ManagerReserveWidget::restore_doc(){
    if(oldmodel != 0){
        delete ui->table_widget->model();
        ui->table_widget->setModel(oldmodel);
        ui->table_widget->resizeColumnsToContents();
    }
}


//////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////// –ö–õ–ê–°–° ///////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////// –î–û–ë–ê–í–ò–¢–¨ –ü–û–ó–ò–¶–ò–Æ ////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////

ReserveAddPosition::ReserveAddPosition(int id, int num, int quantity, double retail_price, double whole_price, int whole_begin, QWidget *parent) : QDialog(parent)
{
    editing = false;

    // —Å–æ—Ö—Ä–∞–Ω—è–µ–º –∫–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä
    grey_id = id;
    doc_num = num;
    base_quantity = quantity;
    base_retail_price = retail_price;
    base_whole_price = whole_price;
    base_whole_begin = whole_begin;

    // —Å–æ–∑–¥–∞—ë–º –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å
    quantity_total = new QLineEdit();
    quantity_spin = new QLineEdit();
    //base_price = new QLineEdit();
    price = new QDoubleSpinBox();
    note = new QLineEdit();
    accept_button = new QPushButton("–ó–∞–Ω–µ—Å—Ç–∏");
    reject_button = new QPushButton("–û—Ç–º–µ–Ω–∏—Ç—å");

    price->setMaximum(1000000);
    price->setSuffix(" —Ä.");
    QIntValidator *intvalid = new QIntValidator(0, base_quantity, this);
    quantity_spin->setValidator(intvalid);
    quantity_total->setEnabled(false);
    //base_price->setEnabled(false);

    quantity_total->setText(QString("%1 —à—Ç.").arg(base_quantity));
    //quantity_spin->setText("0");
    set_base_price_text(base_retail_price);
    price->setValue(base_retail_price);

    set_layout();
    this->adjustSize();
    this->setFixedSize(this->size());
    this->setWindowTitle("–í—ã–ø–∏—Å–∞—Ç—å");

    price_changed = false;  // –µ—Å–ª–∏ —Ü–µ–Ω–∞ —Ö–æ—Ç—å —Ä–∞–∑ –∏–∑–º–µ–Ω–µ–Ω–∞ –≤—Ä—É—á–Ω—É—é, –¥–µ–ª–∞–µ–º —ç—Ç–æ true –∏ –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏ –µ—ë –±–æ–ª–µ –Ω–µ –º–µ–Ω—è–µ–º

    QObject::connect(this->price, SIGNAL(editingFinished()), SLOT(price_entered()));
    QObject::connect(this->quantity_spin, SIGNAL(valueChanged(int)), SLOT(quantity_changed(int)));

    QObject::connect(this->accept_button, SIGNAL(clicked()), SLOT(accept()));
    QObject::connect(this->reject_button, SIGNAL(clicked()), SLOT(reject()));
}

void ReserveAddPosition::set_layout(){
    QGridLayout *grid_lt = new QGridLayout(this);

    // –ª–µ–π–±–ª—ã –¥–ª—è –ø–æ–¥–ø–∏—Å–µ–π
    QLabel *qtotal_label = new QLabel("–í –Ω–∞–ª–∏—á–∏–∏");
    QLabel *bprice_label = new QLabel("–ë–∞–∑. —Ü–µ–Ω–∞");
    QLabel *quant_label = new QLabel("–í—ã–ø–∏—Å–∞—Ç—å");
    QLabel *price_label = new QLabel("–¶–µ–Ω–∞");
    QLabel *note_label = new QLabel("–¢—Ä–µ–±–æ–≤–∞–Ω–∏–µ");

    QLabel *price_retail_label = new QLabel(QString("– –æ–∑–Ω.: %1 —Ä.").arg(base_retail_price));
    QLabel *price_whole_label = new QLabel(QString("–û–ø—Ç.: %1 —Ä.").arg(base_whole_price));
    QLabel *price_begin_label = new QLabel(QString("–û–ø—Ç —Å %1 —à—Ç.").arg(base_whole_begin));
    QLayout *prices_lt = new QVBoxLayout();
    prices_lt->addWidget(price_retail_label);
    prices_lt->addWidget(price_whole_label);
    prices_lt->addWidget(price_begin_label);


    grid_lt->addWidget(qtotal_label, 0, 0);
    grid_lt->addWidget(quantity_total, 0, 1);
    grid_lt->addWidget(bprice_label, 0, 2);
    //grid_lt->addWidget(base_price, 0, 3);
    grid_lt->addLayout(prices_lt, 0, 3);

    grid_lt->addWidget(quant_label, 1, 0);
    grid_lt->addWidget(quantity_spin, 1, 1);
    grid_lt->addWidget(price_label, 1, 2);
    grid_lt->addWidget(price, 1, 3);

    grid_lt->addWidget(note_label, 2, 0);
    grid_lt->addWidget(note, 2, 1, 1, 3);

    // –¥–ª—è –∫–Ω–æ–ø–æ–∫ —Å–¥–µ–ª–∞–µ–º –æ—Ç–¥–µ–ª—å–Ω—ã–π –ª–∞–π–∞—É—Ç
    QHBoxLayout *b_lt = new QHBoxLayout();
    b_lt->addWidget(accept_button);
    b_lt->addWidget(reject_button);

    grid_lt->addLayout(b_lt, 3, 0, 1, 4);

    this->setLayout(grid_lt);
    this->setFixedWidth(350);
}

void ReserveAddPosition::set_base_price_text(double p){
    //base_price->setText(QString("%1 —Ä.").arg(QString::number(p, 'f', 2)));
}

void ReserveAddPosition::accept(){
    if(!editing){
        if(quantity_spin->text().toInt() > base_quantity){
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥—ë–Ω–Ω–æ–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –±–æ–ª—å—à–µ, —á–µ–º –Ω–∞ —Å–∫–ª–∞–¥–µ", QMessageBox::Ok);
            return;
        }
        if(quantity_spin->text().toInt() <= 0){
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥—ë–Ω–Ω–æ–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –Ω—É–ª–µ–≤–æ–µ", QMessageBox::Ok);
            return;
        }
        if(price->value() < 0){
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–¶–µ–Ω–∞ –Ω–µ –º–æ–∂–µ—Ç –±—ã—Ç—å –æ—Ç—Ä–∏—Ü–∞—Ç–µ–ª—å–Ω–æ–π!", QMessageBox::Ok);
            return;
        }
        if(quantity_spin->text().toInt() == 0){
            QDialog::reject();
            return;
        }

        // —Å–æ—Ö—Ä–∞–Ω–∏—Ç—å –≤ –±–∞–∑—É, –≤ —Ç–∞–±–ª–∏—Ü—É reserve
        // –µ—Å–ª–∏ —Ç–∞–∫–æ–π id —É–∂–µ –µ—Å—Ç—å –≤ —ç—Ç–æ–π –∑–∞—è–≤–∫–µ - –¥–æ–±–∞–≤–ª—è–µ–º –∫–æ–ª-–≤–æ, –∏–Ω–∞—á–µ —Å–æ–∑–¥–∞—ë–º –Ω–æ–≤—É—é –∑–∞–ø–∏—Å—å
        QSqlQuery query(base);
        QString query_str = QString("SELECT count(*) FROM reserve WHERE grey_id = %1 AND doc_id = %2")
                .arg(grey_id)
                .arg(QString("(SELECT id FROM reserve_docs WHERE doc_num = %1 AND user_id = %2)").arg(doc_num).arg(USER_ID));
        if(!query.exec(query_str)){
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", QString("–ù–µ —É–¥–∞–ª–æ—Å—å: %1").arg(query.lastError().text()), QMessageBox::Ok);
            return;
        }
        query.next();
        if(query.value(0).toInt() == 1){
            query_str = QString("UPDATE reserve SET quantity = quantity + %1 WHERE grey_id = %2 AND doc_id = %3")
                    .arg(quantity_spin->text())
                    .arg(grey_id)
                    .arg(QString("(SELECT id FROM reserve_docs WHERE doc_num = %1 AND user_id = %2)").arg(doc_num).arg(USER_ID));
            if(!query.exec(query_str)){
                QMessageBox::warning(this, "–û—à–∏–±–∫–∞", QString("–ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å –∫ –∫–æ–ª–∏—á–µ—Å—Ç–≤—É: %1").arg(query.lastError().text()), QMessageBox::Ok);
                return;
            }
        }
        else{
            query_str = QString("INSERT INTO reserve(doc_id, grey_id, quantity, price, note) "
                                        "VALUES(%1, %2, %3, %4, '%5')")
                    .arg(QString("(SELECT id FROM reserve_docs WHERE doc_num = %1 AND user_id = %2)").arg(doc_num).arg(USER_ID))
                    .arg(grey_id).arg(quantity_spin->text()).arg(price->value()).arg(note->text());
            qDebug() << query_str;
            if(!query.exec(query_str)){
                QMessageBox::warning(this, "–û—à–∏–±–∫–∞", QString("–ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å –≤ –∑–∞—è–≤–∫—É: %1").arg(query.lastError().text()), QMessageBox::Ok);
                return;
            }
        }
    }
    QDialog::accept();
}


void ReserveAddPosition::price_entered(){
    if(!price_changed){
        qDebug("–¶–µ–Ω–∞ –≤–≤–µ–¥–µ–Ω–∞ –≤—Ä—É—á–Ω—É—é.");
        price_changed = true;
    }
}

void ReserveAddPosition::quantity_changed(int q){
    //set_base_price_text(q < base_whole_begin ? base_retail_price : base_whole_price);
    //if(!price_changed)
    //    price->setValue(q < base_whole_begin ? base_retail_price : base_whole_price);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////// –ö–õ–ê–°–° ///////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////// –í–´–ß–ï–°–¢–¨ –ö–û–õ–ò–ß–ï–°–¢–í–û ///////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////

ReserveSubtract::ReserveSubtract(int id, int quantity, QWidget *parent){
    c_id = id;
    c_quantity = quantity;

    subtr_spin = new QSpinBox();
    quant_line = new QLineEdit();
    accept_button = new QPushButton("–í—ã—á–µ—Å—Ç—å");
    reject_button = new QPushButton("–û—Ç–º–µ–Ω–∞");

    QLabel *s_label = new QLabel("–í—ã—á–µ—Å—Ç—å");
    QLabel *q_label = new QLabel("–ö–æ–ª-–≤–æ");

    subtr_spin->setValue(0);
    subtr_spin->setMinimum(0);
    subtr_spin->setMaximum(c_quantity);
    quant_line->setText(QString("%1 —à—Ç.").arg(c_quantity));

    QGridLayout *lt = new QGridLayout(this);
    QVBoxLayout *vlt1 = new QVBoxLayout();
    QVBoxLayout *vlt2 = new QVBoxLayout();
    vlt1->addWidget(s_label);
    vlt1->addWidget(subtr_spin);
    vlt2->addWidget(q_label);
    vlt2->addWidget(quant_line);
    lt->addLayout(vlt1, 0, 0);
    lt->addLayout(vlt2, 0, 1);
    lt->addWidget(accept_button, 1, 0);
    lt->addWidget(reject_button, 1, 1);

    this->setWindowTitle("–í—ã—á–µ—Å—Ç—å");

    connect(subtr_spin, SIGNAL(valueChanged(int)), SLOT(change_quant(int)));
    connect(accept_button, SIGNAL(clicked()), SLOT(accept()));
    connect(reject_button, SIGNAL(clicked()), SLOT(reject()));
}

void ReserveSubtract::change_quant(int q){
    quant_line->setText(QString("%1 —à—Ç.").arg(c_quantity - q));
}

void ReserveSubtract::accept(){
    if(subtr_spin->value() > 0){
        QSqlQuery query(base);
        QString query_str = QString("UPDATE reserve SET quantity = quantity - %1 WHERE id = %2").arg(subtr_spin->value()).arg(c_id);
        if(!query.exec(query_str)){
            QMessageBox::warning(this, "–ù–µ —É–¥–∞–ª–æ—Å—å –≤—ã—á–µ—Å—Ç—å", query.lastError().text(), QMessageBox::Ok);
            return;
        }
    }
    QDialog::accept();
}


//////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////// –ö–õ–ê–°–° ///////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////// –ü–ï– –ï–ú–ï–°–¢–ò–¢–¨ –ü–û–ó–ò–¶–ò–ò //////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////

ReserveMovePositions::ReserveMovePositions(QTableView *table, QModelIndexList indexes, QWidget *parent) : QDialog(parent){
    create_widgets();
    set_layout();

    max_docs = 10;
    reserves_button->setChecked(true);
    form_reserves_list();

    switch_widget();

    connects();
}

inline void ReserveMovePositions::create_widgets(){
    reserves_button = new QRadioButton("–ó–∞—è–≤–∫–∏");
    docs_button = new QRadioButton("–î–æ–∫—É–º–µ–Ω—Ç—ã");
    offers_button = new QRadioButton("–ö–æ–º–º. –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è");
    ok_button = new QPushButton("–ü–µ—Ä–µ–º–µ—Å—Ç–∏—Ç—å");
    cancel_button = new QPushButton("–û—Ç–º–µ–Ω–∞");
    layout = new QStackedLayout();
    docs_widget = new QWidget(this);
    offers_widget = new QWidget(this);
    reserves_list = new QListView();
    docs_list = new QTableView();
    offers_list = new QTableView();


}

inline void ReserveMovePositions::set_layout(){
    // –ª–∞–π–∞—É—Ç —Å —Ç—Ä–µ–º—è —Ä–∞–¥–∏–æ–∫–Ω–æ–ø–∫–∞–º–∏, –æ–Ω–∏ –±—É–¥—É—Ç —Ä–∞—Å–ø–æ–ª–æ–∂–µ–Ω—ã –≤–µ—Ä—Ç–∏–∫–∞–ª—å–Ω–æ –∏ —Å–¥–≤–∏–Ω—É—Ç—ã –∫–≤–µ—Ä—Ö—É
    QVBoxLayout *radiobuttons_vlt = new QVBoxLayout();
    radiobuttons_vlt->addWidget(reserves_button);
    radiobuttons_vlt->addWidget(docs_button);
    radiobuttons_vlt->addWidget(offers_button);
    radiobuttons_vlt->addStretch(1);

    // –ª–∞–π–∞—É—Ç —Å –∫–Ω–æ–ø–∫–∞–º–∏ –û–ö –∏ –û—Ç–º–µ–Ω–∞, –æ–Ω–∏ –±—É–¥—É—Ç –≤ —Å–∞–º–æ–º –Ω–∏–∑—É —Å–ø—Ä–∞–≤–∞
    QHBoxLayout *buttons_hlt = new QHBoxLayout();
    buttons_hlt->addStretch(1);
    buttons_hlt->addWidget(ok_button);
    buttons_hlt->addWidget(cancel_button);

    // —Ñ–æ—Ä–º–∏—Ä—É–µ–º docs_list
    //

    // —Ñ–æ—Ä–º–∏—Ä—É–µ–º offers_list
    //

    // —Ñ–æ—Ä–º–∏—Ä—É–µ–º stacked_layout
    layout->addWidget(reserves_list);
    layout->addWidget(docs_list);
    layout->addWidget(offers_list);

    // –≥–ª–∞–≤–Ω—ã–π –ª–∞–π–∞—É—Ç
    // —Å–æ—Å—Ç–æ–∏—Ç –∏–∑ –¥–≤—É—Ö: —Å–≤–µ—Ä—Ö—É hbox —Å —Ä–∞–¥–∏–æ–∫–Ω–æ–ø–∫–∞–º–∏ –∏ —Å—Ç–∞–∫–µ–¥–ª–∞–π–∞—É—Ç–æ–º, —Å–Ω–∏–∑—É –∫–Ω–æ–ø–∫–∏ –æ–∫ –∏ –æ—Ç–º–µ–Ω–∞
    QVBoxLayout *main_lt = new QVBoxLayout();
    QHBoxLayout *top_lt = new QHBoxLayout();
    top_lt->addLayout(radiobuttons_vlt);
    top_lt->addLayout(layout);
    main_lt->addLayout(top_lt);
    main_lt->addLayout(buttons_hlt);

    this->setLayout(main_lt);
}

inline void ReserveMovePositions::connects(){
    // –ø–µ—Ä–µ–∫—é—á–µ–Ω–∏–µ –≤–∏–¥–∂–µ—Ç–æ–≤ –ø–æ —Ä–∞–¥–∏–æ–∫–Ω–æ–ø–∫–∞–º
    connect(this->reserves_button, SIGNAL(clicked()), SLOT(switch_widget()));
    connect(this->docs_button, SIGNAL(clicked()), SLOT(switch_widget()));
    connect(this->offers_button, SIGNAL(clicked()), SLOT(switch_widget()));
}

inline void ReserveMovePositions::form_reserves_list(){
    // –≤ —Ç–∞–±–ª–∏—Ü–µ –±—É–¥—É—Ç –¥–≤–∞ —Å—Ç–æ–ª–±—Ü–∞ - "–ó–∞—è–≤–∫–∞ ‚Ññ" –∏ "–î–ª—è –∫–æ–≥–æ"
    // –∫ –∫–∞–∂–¥–æ–π —è—á–µ–π–∫–µ —Å –Ω–æ–º–µ—Ä–æ–º –∑–∞—è–≤–∫–∏ –ø—Ä–∏—Å–æ–≤–æ–∫—É–ø–ª—è–µ–º –æ–Ω—ã–π –≤ UserRole

}

inline void ReserveMovePositions::form_docs_list(){

}

inline void ReserveMovePositions::form_offers_list(){

}



////// –°–õ–û–¢–´
void ReserveMovePositions::switch_widget(){
    if(this->reserves_button->isChecked())
        layout->setCurrentWidget(reserves_list);
    else if(this->docs_button->isChecked())
        layout->setCurrentWidget(docs_list);
    else if(this->offers_button->isChecked())
        layout->setCurrentWidget(offers_list);
}
